lib/smack/SmackRep.cpp

```diff
@@ -181,19 +222,88 @@ const Stmt* SmackRep::memset(const llvm::MemSetInst& msi) {
+  for (unsigned i = 3; i < msi.getNumOperands() - 1; i++)
     args.push_back(expr(msi.getOperand(i)));
   return Stmt::call(name.str(),args);
 }
 
-const Expr* SmackRep::pa(const Expr* base, int index, int size) {
-  return pa(base, Expr::lit(index), Expr::lit(size));
+const Stmt* SmackRep::load(const llvm::Value* addr, const llvm::Value* val) {
+// The tricky part is that we could expr(li) is actually expr(val) so that it is possible to pass li to val. 
+  const Expr* rhs = mem(addr);
+
+  if (isFloat(val))
+    rhs = Expr::fn("$si2fp", rhs);
+
+  if (!SmackOptions::BitVectors || (SmackOptions::InferFieldOverlap && isFieldDisjoint(addr, llvm::cast<const llvm::Instruction>(val))))
+    return Stmt::assign(expr(val), rhs);
+  else {
+    stringstream name;
+    name << "$load." << int_type(getElementSize(addr));
+    return Stmt::assign(expr(val), Expr::fn(name.str(), Expr::id(memPath(getRegion(addr), 8)), expr(addr)));
+  }
+}
+
+const Stmt* SmackRep::store(const llvm::Value* addr, const llvm::Value* val, const llvm::StoreInst* si) {
+// Having a default value of si (NULL) is unsound.
+  const Expr* rhs = expr(val);
+
+  if (isFloat(val))
+    rhs = Expr::fn("$fp2si", rhs);
+
+  if (!SmackOptions::BitVectors || (SmackOptions::InferFieldOverlap && isFieldDisjoint(addr, si)))
+    return Stmt::assign(mem(addr),rhs);
+  else
+    return storeAsBytes(getRegion(addr), getElementSize(addr), expr(addr), expr(val));
+}
+
+const Stmt* SmackRep::storeAsBytes(unsigned region, unsigned size, const Expr* p, const Expr* e)
+{
+  stringstream name;
+  name << "$store." << int_type(size);
+  return Stmt::assign(Expr::id(memPath(region, 8)), Expr::fn(name.str(), Expr::id(memPath(region, 8)), p, e));
+
+}
+
+bool SmackRep::isFieldDisjoint(const llvm::Value* ptr, const llvm::Instruction* inst) 
+{
+  return aliasAnalysis->isFieldDisjoint(ptr, inst); 
+}
+
+bool SmackRep::isFieldDisjoint(const llvm::GlobalValue *V, unsigned offset) 
+{
+  return aliasAnalysis->isFieldDisjoint(V, offset); 
+}
+
+bool SmackRep::isTypeSafe(const llvm::Value* ptr, const llvm::Instruction* inst) 
+{
+  return aliasAnalysis->isTypeSafe(ptr, inst); 
+}
+
+bool SmackRep::isTypeSafe(const llvm::GlobalValue *V) 
+{
+  return aliasAnalysis->isTypeSafe(V); 
+}
+
+const Expr* SmackRep::pa(const Expr* base, int index, int size, int i_size, int t_size) {
+    return pa(base, lit(index, i_size), lit(size, t_size), i_size, t_size);
 }
-const Expr* SmackRep::pa(const Expr* base, const Expr* index, int size) {
-  return pa(base, index, Expr::lit(size));
+const Expr* SmackRep::pa(const Expr* base, const Expr* index, int size, int i_size, int t_size) {
+  return pa(base, index, lit(size, t_size), i_size, t_size);
 }
-const Expr* SmackRep::pa(const Expr* base, const Expr* index, const Expr* size) {
-  return Expr::fn("$pa", base, index, size);
+const Expr* SmackRep::pa(const Expr* base, const Expr* index, const Expr* size, int i_size, int t_size) {
+    if (i_size == 32) 
```

Comment:
This could be cleaner and more robust. Something like this:
````
if (i_size < t_size)
  index = Expr::fn("$zext.XX.YY", index);
if (i_size > t_size)
  size = Expr::fn("$zext.YY.XX", size);
Expr* offset = Expr::fn("$mul.ZZ", index, size);
if (ZZ < width)
  offset = Expr::fn("$zext.ZZ.ref", offset);
return Expr::fn("$add.ref", base, offset);
````
where `XX == i_size`, `YY == t_size`, and `ZZ == max(XX,YY)`. Of course you'll have to make a few changes (e.g. because `index` and `size` are `const` pointers, and `XX`, `YY`, `ZZ` must be injected into the strings), but you get the idea.

Also, the `i_size` and `t_size` arguments should be unsigned integers.